File: src/Sample/Plumix.Sample/Demos/Material/SelectionHandlesDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Plumix.Gestures;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

// Dart parity source: dart_sample/lib/demos/material/selection_handles_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class SelectionHandlesDemoPage : StatefulWidget
{
    public override State CreateState() => new SelectionHandlesDemoPageState();
}

internal sealed class SelectionHandlesDemoPageState : State
{
    private const double LineTop = 96;
    private const double LineHeight = 24;

    private readonly LayerLink _startHandleLink = new();
    private readonly LayerLink _endHandleLink = new();
    private readonly LayerLink _toolbarLink = new();
    private readonly TextEditingController _fieldController = new(
        "Long press this real text field, then drag either selection handle.");

    private SelectionOverlay? _overlay;
    private double _startX = 48;
    private double _endX = 232;
    private bool _handlesVisible;
    private bool _collapsed;

    public override void Dispose()
    {
        _overlay?.Dispose();
        _overlay = null;
        _fieldController.Dispose();
    }

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 12,
            children:
            [
                new Text("SelectionOverlay + Material handles", fontSize: 20, color: Colors.Black),
                new Text(
                    "Drag either handle to move its endpoint. Collapsed mode keeps a single upward handle.",
                    fontSize: 14,
                    color: Color.Parse("#8A000000")),
                new TextField(
                    controller: _fieldController,
                    maxLines: 2,
                    decoration: new InputDecoration(
                        labelText: "RenderEditable-backed handles",
                        border: new OutlineInputBorder())),
                new Row(
                    spacing: 8,
                    children:
                    [
                        ControlButton(_handlesVisible ? "Hide handles" : "Show handles", ToggleHandles),
                        ControlButton(_collapsed ? "Ranged" : "Collapsed", ToggleCollapsed),
                        ControlButton("Reset", ResetEndpoints),
                    ]),
                new Expanded(
                    child: new Container(
                        color: Color.Parse("#FFF7F2FA"),
                        child: new Stack(
                            clipBehavior: Clip.None,
                            children:
                            [
                                new Positioned(
                                    left: 24,
                                    top: LineTop - LineHeight,
                                    right: 24,
                                    child: new Text(
                                        "Drag the handles across this line",
                                        fontSize: 18,
                                        color: Color.Parse("#FF1D192B"))),
                                new Positioned(
                                    left: EffectiveStartX,
                                    top: LineTop,
                                    child: new CompositedTransformTarget(
                                        link: _startHandleLink,
                                        child: new SizedBox(width: 0, height: 0))),
                                new Positioned(
                                    left: _endX,
                                    top: LineTop,
                                    child: new CompositedTransformTarget(
                                        link: _endHandleLink,
                                        child: new SizedBox(width: 0, height: 0))),
                                new Positioned(
                                    left: 24,
                                    top: LineTop + 24,
                                    child: new CompositedTransformTarget(
                                        link: _toolbarLink,
                                        child: new SizedBox(width: 0, height: 0))),
                                new Positioned(
                                    left: 24,
                                    bottom: 18,
                                    child: new Text(
                                        $"startX={EffectiveStartX:0}, endX={_endX:0}, "
                                        + $"handles={(_handlesVisible ? "on" : "off")}",
                                        fontSize: 12,
                                        color: Color.Parse("#FF6750A4"))),
                            ]))),
            ]);
    }

    private double EffectiveStartX => _collapsed ? _endX : _startX;

    private void ToggleHandles()
    {
        SelectionOverlay overlay = EnsureOverlay();
        if (_handlesVisible)
        {
            overlay.HideHandles();
        }
        else
        {
            overlay.ShowHandles();
        }

        SetState(() => _handlesVisible = !_handlesVisible);
    }

    private void ToggleCollapsed()
    {
        SetState(() => _collapsed = !_collapsed);
        SyncOverlay();
    }

    private void ResetEndpoints()
    {
        SetState(() =>
        {
            _startX = 48;
            _endX = 232;
        });
        SyncOverlay();
    }

    private SelectionOverlay EnsureOverlay()
    {
        if (_overlay is not null)
        {
            return _overlay;
        }

        _overlay = new SelectionOverlay(
            context: Context,
            startHandleType: TextSelectionHandleType.Left,
            lineHeightAtStart: LineHeight,
            endHandleType: TextSelectionHandleType.Right,
            lineHeightAtEnd: LineHeight,
            selectionEndpoints: BuildEndpoints(),
            selectionControls: MaterialTextSelectionHandleControls.Instance,
            selectionDelegate: null,
            clipboardStatus: null,
            startHandleLayerLink: _startHandleLink,
            endHandleLayerLink: _endHandleLink,
            toolbarLayerLink: _toolbarLink,
            onStartHandleDragUpdate: details => MoveStart(details.Delta.X),
            onEndHandleDragUpdate: details => MoveEnd(details.Delta.X));
        SyncOverlay();
        return _overlay;
    }

    private void MoveStart(double delta)
    {
        SetState(() => _startX = Math.Clamp(_startX + delta, 24, _endX));
        SyncOverlay();
    }

    private void MoveEnd(double delta)
    {
        SetState(() => _endX = Math.Clamp(_endX + delta, EffectiveStartX, 320));
        SyncOverlay();
    }

    private void SyncOverlay()
    {
        if (_overlay is null)
        {
            return;
        }

        _overlay.StartHandleType = _collapsed
            ? TextSelectionHandleType.Collapsed
            : TextSelectionHandleType.Left;
        _overlay.EndHandleType = _collapsed
            ? TextSelectionHandleType.Collapsed
            : TextSelectionHandleType.Right;
        _overlay.SelectionEndpoints = BuildEndpoints();
        _overlay.MarkNeedsBuild();
    }

    private TextSelectionPoint[] BuildEndpoints()
    {
        return
        [
            new TextSelectionPoint(new Point(EffectiveStartX, LineTop), TextDirection.Ltr),
            new TextSelectionPoint(new Point(_endX, LineTop), TextDirection.Ltr),
        ];
    }

    private static Widget ControlButton(string label, Action onPressed)
    {
        return new TextButton(
            onPressed: onPressed,
            backgroundColor: Color.Parse("#FFEADDFF"),
            foregroundColor: Color.Parse("#FF21005D"),
            minHeight: 36,
            child: new Text(label, fontSize: 12));
    }
}

File: dart_sample/lib/demos/material/selection_handles_demo_page.dart
import 'package:flutter/material.dart';

class SelectionHandlesDemoPage extends StatefulWidget {
  const SelectionHandlesDemoPage({super.key});

  @override
  State<SelectionHandlesDemoPage> createState() =>
      _SelectionHandlesDemoPageState();
}

class _SelectionHandlesDemoPageState extends State<SelectionHandlesDemoPage> {
  static const double _lineTop = 96;
  static const double _lineHeight = 24;

  final LayerLink _startHandleLink = LayerLink();
  final LayerLink _endHandleLink = LayerLink();
  final LayerLink _toolbarLink = LayerLink();
  final TextEditingController _fieldController = TextEditingController(
    text: 'Long press this real text field, then drag either selection handle.',
  );

  SelectionOverlay? _overlay;
  double _startX = 48;
  double _endX = 232;
  bool _handlesVisible = false;
  bool _collapsed = false;

  @override
  void dispose() {
    _overlay?.dispose();
    _overlay = null;
    _fieldController.dispose();
    super.dispose();
  }

  double get _effectiveStartX => _collapsed ? _endX : _startX;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: <Widget>[
        const Text(
          'SelectionOverlay + Material handles',
          style: TextStyle(fontSize: 20),
        ),
        const SizedBox(height: 12),
        const Text(
          'Drag either handle to move its endpoint. Collapsed mode keeps a single upward handle.',
          style: TextStyle(fontSize: 14, color: Color(0x8A000000)),
        ),
        const SizedBox(height: 12),
        TextField(
          controller: _fieldController,
          maxLines: 2,
          decoration: const InputDecoration(
            labelText: 'RenderEditable-backed handles',
            border: OutlineInputBorder(),
          ),
        ),
        const SizedBox(height: 12),
        Row(
          children: <Widget>[
            _controlButton(
              _handlesVisible ? 'Hide handles' : 'Show handles',
              _toggleHandles,
            ),
            const SizedBox(width: 8),
            _controlButton(_collapsed ? 'Ranged' : 'Collapsed', _toggleCollapsed),
            const SizedBox(width: 8),
            _controlButton('Reset', _resetEndpoints),
          ],
        ),
        const SizedBox(height: 12),
        Expanded(
          child: ColoredBox(
            color: const Color(0xFFF7F2FA),
            child: Stack(
              clipBehavior: Clip.none,
              children: <Widget>[
                const Positioned(
                  left: 24,
                  top: _lineTop - _lineHeight,
                  right: 24,
                  child: Text(
                    'Drag the handles across this line',
                    style: TextStyle(fontSize: 18, color: Color(0xFF1D192B)),
                  ),
                ),
                Positioned(
                  left: _effectiveStartX,
                  top: _lineTop,
                  child: CompositedTransformTarget(
                    link: _startHandleLink,
                    child: const SizedBox.shrink(),
                  ),
                ),
                Positioned(
                  left: _endX,
                  top: _lineTop,
                  child: CompositedTransformTarget(
                    link: _endHandleLink,
                    child: const SizedBox.shrink(),
                  ),
                ),
                Positioned(
                  left: 24,
                  top: _lineTop + 24,
                  child: CompositedTransformTarget(
                    link: _toolbarLink,
                    child: const SizedBox.shrink(),
                  ),
                ),
                Positioned(
                  left: 24,
                  bottom: 18,
                  child: Text(
                    'startX=${_effectiveStartX.toStringAsFixed(0)}, '
                    'endX=${_endX.toStringAsFixed(0)}, '
                    'handles=${_handlesVisible ? 'on' : 'off'}',
                    style: const TextStyle(
                      fontSize: 12,
                      color: Color(0xFF6750A4),
                    ),
                  ),
                ),
              ],
            ),
          ),
        ),
      ],
    );
  }

  void _toggleHandles() {
    final SelectionOverlay overlay = _ensureOverlay();
    if (_handlesVisible) {
      overlay.hideHandles();
    } else {
      overlay.showHandles();
    }
    setState(() => _handlesVisible = !_handlesVisible);
  }

  void _toggleCollapsed() {
    setState(() => _collapsed = !_collapsed);
    _syncOverlay();
  }

  void _resetEndpoints() {
    setState(() {
      _startX = 48;
      _endX = 232;
    });
    _syncOverlay();
  }

  SelectionOverlay _ensureOverlay() {
    final SelectionOverlay? existing = _overlay;
    if (existing != null) {
      return existing;
    }

    final SelectionOverlay overlay = SelectionOverlay(
      context: context,
      startHandleType: TextSelectionHandleType.left,
      lineHeightAtStart: _lineHeight,
      endHandleType: TextSelectionHandleType.right,
      lineHeightAtEnd: _lineHeight,
      selectionEndpoints: _buildEndpoints(),
      selectionControls: materialTextSelectionHandleControls,
      selectionDelegate: null,
      clipboardStatus: null,
      startHandleLayerLink: _startHandleLink,
      endHandleLayerLink: _endHandleLink,
      toolbarLayerLink: _toolbarLink,
      onStartHandleDragUpdate: (DragUpdateDetails details) =>
          _moveStart(details.delta.dx),
      onEndHandleDragUpdate: (DragUpdateDetails details) =>
          _moveEnd(details.delta.dx),
    );
    _overlay = overlay;
    _syncOverlay();
    return overlay;
  }

  void _moveStart(double delta) {
    setState(() => _startX = (_startX + delta).clamp(24, _endX));
    _syncOverlay();
  }

  void _moveEnd(double delta) {
    setState(() => _endX = (_endX + delta).clamp(_effectiveStartX, 320));
    _syncOverlay();
  }

  void _syncOverlay() {
    final SelectionOverlay? overlay = _overlay;
    if (overlay == null) {
      return;
    }

    overlay
      ..startHandleType = _collapsed
          ? TextSelectionHandleType.collapsed
          : TextSelectionHandleType.left
      ..endHandleType = _collapsed
          ? TextSelectionHandleType.collapsed
          : TextSelectionHandleType.right
      ..selectionEndpoints = _buildEndpoints()
      ..markNeedsBuild();
  }

  List<TextSelectionPoint> _buildEndpoints() {
    return <TextSelectionPoint>[
      TextSelectionPoint(
        Offset(_effectiveStartX, _lineTop),
        TextDirection.ltr,
      ),
      TextSelectionPoint(Offset(_endX, _lineTop), TextDirection.ltr),
    ];
  }

  Widget _controlButton(String label, VoidCallback onPressed) {
    return TextButton(
      onPressed: onPressed,
      style: TextButton.styleFrom(
        backgroundColor: const Color(0xFFEADDFF),
        foregroundColor: const Color(0xFF21005D),
        minimumSize: const Size(0, 36),
      ),
      child: Text(label, style: const TextStyle(fontSize: 12)),
    );
  }
}
